File: app/src/main/java/com/kdroid/kotlintuts/geeks/recursion/Factorial.java
package com.kdroid.kotlintuts.geeks.recursion;

public class Factorial {
    public static void main(String[] args) {
        int n = 9;
        System.out.println(getFactorial(5));
    }

    public static int getFactorial(int n) {
        if(n==0|| n==1){
            return 1;
        }else {
            return n*getFactorial(n-1);
        }
    }

}

File: app/src/main/java/com/kdroid/kotlintuts/codingchallanges/geeks/recursion/Factorial.kt
package com.kdroid.kotlintuts.codingchallanges.geeks.recursion

import java.math.BigInteger

fun main(args: Array<String>) {
    val number = 140L
    print(getFactorialNumber(number, 1))
}

tailrec fun getFactorialNumber(n: Long, accu: Long = 1): Long {
    var soFar = accu * n
    return if (n <=1) {
        soFar
    } else {
        getFactorialNumber(n - 1, soFar)
    }
}


tailrec fun getFibonacciNumber(n: Int, a: BigInteger, b: BigInteger): BigInteger {

    if (n == 0) {
        return b;
    } else {
        print(" $a ")
        return getFibonacciNumber(n - 1, a + b, a);
    }
}